consensus/state.go

```diff
@@ -1412,9 +1428,9 @@ func (cs *ConsensusState) addVote(vote *types.Vote, peerID p2p.ID) (added bool,
 			}
 		}
 		// Update ValidBlock
-		if ok && !blockID.IsZero() && !cs.ValidBlock.HashesTo(blockID.Hash) && vote.Round > cs.ValidRound {
+		if ok && !blockID.IsZero() && (cs.ValidRound < vote.Round) && (vote.Round <= cs.Round) {
 			// update valid value
-			if cs.ProposalBlock.HashesTo(blockID.Hash) {
+			if !cs.ValidBlock.HashesTo(blockID.Hash) && cs.ProposalBlock.HashesTo(blockID.Hash) {
```

Comment:
With this update it looks like ValidRound won't properly update, and we might end up with a stale (lower) ValidRound and perhaps even regress to an invalid ValidBlock later (maybe).  So I think setting cs.ValidRound = vote.Round even when cs.ValidBlock.HashesTo(blockID.Hash) would be better, in case we we're tracking a lower prevote set (e.g. for a proposer's POL-round prevote gossip) and that round happens to be less than this vote.Round.

Reply:
That's good point, we should update also when cs.ValidBlock.HashesTo(blockID.Hash) as we now have a PoLK from a more recent round for the same block. I fixed this.